File: services/api/src/__tests__/reminder-recurrence.test.ts
/**
 * NOVA API — the stored repeat rule and when it next comes due.
 *
 * §14 of the acceptance criteria requires "Remind me every Monday" to work, and
 * the model used to refuse it: *"I can only set it for a specific date and time
 * rather than a recurring schedule."* `reminders.repeat_rule` existed as a column
 * and the REST route accepted it, but nothing read it, so the honest answer was
 * the one the model gave.
 *
 * This file pins the two halves that are easy to get wrong and invisible in a
 * manual test: the **grammar** (a rule is either exactly one of three forms or it
 * is refused, with a message the model can act on) and the **arithmetic** —
 * month ends, a weekly rule across a month boundary, wall-clock times preserved
 * across a DST change, and the strict "after" boundary.
 *
 * IST is a trap here and is deliberately not the only zone under test: it has no
 * DST and a :30 offset, so code that adds 24 hours per "day" and code that adds
 * one calendar day agree, and code that ignores offsets by an hour still looks
 * right at :00. Europe/London and a 00:30 local time are what separate them.
 */
import { describe, expect, it } from 'vitest';
import {
	MAX_REPEAT_INTERVAL,
	describeRepeatRule,
	formatRepeatRule,
	nextOccurrence,
	parseRepeatRule,
} from '../services/reminder-recurrence.js';

/** Parses a rule the test knows is valid, failing loudly if it is not. */
function rule(text: string) {
	const parsed = parseRepeatRule(text);
	if (!parsed.ok) throw new Error(`fixture rule ${text} did not parse: ${parsed.error}`);
	return parsed.rule;
}

/** The next occurrence after [after] for [text], as an ISO instant. */
function next(text: string, start: string, after: string, timezone = 'Asia/Kolkata'): string {
	return nextOccurrence(rule(text), new Date(start), new Date(after), timezone).toISOString();
}

describe('parseRepeatRule — exactly one of three forms, or a clear refusal', () => {
	it.each([
		['FREQ=DAILY', { frequency: 'DAILY', interval: 1 }],
		['FREQ=WEEKLY;BYDAY=MO', { frequency: 'WEEKLY', interval: 1, byDay: 'MO' }],
		['FREQ=MONTHLY;BYMONTHDAY=15', { frequency: 'MONTHLY', interval: 1, byMonthDay: 15 }],
		['FREQ=WEEKLY;INTERVAL=2;BYDAY=SU', { frequency: 'WEEKLY', interval: 2, byDay: 'SU' }],
		['FREQ=MONTHLY;INTERVAL=3;BYMONTHDAY=1', { frequency: 'MONTHLY', interval: 3, byMonthDay: 1 }],
	])('accepts %s', (text, expected) => {
		const parsed = parseRepeatRule(text);
		expect(parsed.ok).toBe(true);
		if (!parsed.ok) return;
		expect(parsed.rule).toMatchObject(expected);
	});

	it('accepts lower case and normalises it to the canonical spelling', () => {
		// The model is not a formatter. Rejecting `freq=weekly;byday=mo` would be a
		// refusal about punctuation, and the stored rule is rewritten canonically
		// before it reaches the column, so nothing downstream has to care.
		const parsed = parseRepeatRule('freq=weekly;byday=mo');
		expect(parsed.ok).toBe(true);
		if (!parsed.ok) return;
		expect(formatRepeatRule(parsed.rule)).toBe('FREQ=WEEKLY;BYDAY=MO');
	});

	it('drops the redundant INTERVAL=1 when it writes the rule back out', () => {
		const parsed = parseRepeatRule('FREQ=DAILY;INTERVAL=1');
		expect(parsed.ok).toBe(true);
		if (!parsed.ok) return;
		expect(parsed.rule.interval).toBe(1);
		expect(formatRepeatRule(parsed.rule)).toBe('FREQ=DAILY');
	});

	it.each([
		['', /empty|missing/i],
		['   ', /empty|missing/i],
		['DAILY', /FREQ/i],
		['FREQ=', /FREQ/i],
		['FREQ=YEARLY', /DAILY|WEEKLY|MONTHLY/],
		['FREQ=HOURLY', /DAILY|WEEKLY|MONTHLY/],
		// A weekly rule that names no weekday cannot be resolved to an instant.
		['FREQ=WEEKLY', /BYDAY/i],
		// A daily rule that names one is a contradiction, not a stronger rule.
		['FREQ=DAILY;BYDAY=MO', /BYDAY/i],
		['FREQ=MONTHLY', /BYMONTHDAY/i],
		['FREQ=MONTHLY;BYMONTHDAY=0', /BYMONTHDAY/i],
		['FREQ=MONTHLY;BYMONTHDAY=32', /BYMONTHDAY/i],
		['FREQ=MONTHLY;BYMONTHDAY=abc', /BYMONTHDAY/i],
		['FREQ=WEEKLY;BYDAY=XX', /BYDAY/i],
		['FREQ=WEEKLY;BYDAY=MONDAY', /BYDAY/i],
		['FREQ=DAILY;INTERVAL=0', /INTERVAL/i],
		['FREQ=DAILY;INTERVAL=-1', /INTERVAL/i],
		['FREQ=DAILY;INTERVAL=1.5', /INTERVAL/i],
		[`FREQ=DAILY;INTERVAL=${MAX_REPEAT_INTERVAL + 1}`, /INTERVAL/i],
		// Duplicated parts are ambiguous: which one wins is a guess.
		['FREQ=DAILY;FREQ=WEEKLY', /more than once/i],
		['FREQ=DAILY;INTERVAL=2;INTERVAL=3', /more than once/i],
		// The RFC 5545 prefix is not part of this vocabulary.
		['RRULE:FREQ=DAILY', /FREQ/i],
		['FREQ=DAILY;COUNT=5', /COUNT|unexpected/i],
	])('refuses %j with a message that names the offending part', (text, pattern) => {
		const parsed = parseRepeatRule(text);
		expect(parsed.ok).toBe(false);
		if (parsed.ok) return;
		expect(parsed.error).toMatch(pattern);
	});

	it('refuses a rule too long to be one of the three forms', () => {
		const parsed = parseRepeatRule(`FREQ=DAILY;INTERVAL=${'1'.repeat(200)}`);
		expect(parsed.ok).toBe(false);
		if (parsed.ok) return;
		expect(parsed.error).toMatch(/long|length|120/i);
	});
});

describe('describeRepeatRule — the words the confirmation says out loud', () => {
	it.each([
		['FREQ=DAILY', 'every day'],
		['FREQ=DAILY;INTERVAL=3', 'every 3 days'],
		['FREQ=WEEKLY;BYDAY=MO', 'every Monday'],
		['FREQ=WEEKLY;INTERVAL=2;BYDAY=SU', 'every 2 weeks on Sunday'],
		['FREQ=MONTHLY;BYMONTHDAY=1', 'every month on the 1st'],
		['FREQ=MONTHLY;BYMONTHDAY=15', 'every month on the 15th'],
		['FREQ=MONTHLY;BYMONTHDAY=22', 'every month on the 22nd'],
		['FREQ=MONTHLY;BYMONTHDAY=31', 'every month on the 31st'],
		['FREQ=MONTHLY;INTERVAL=3;BYMONTHDAY=31', 'every 3 months on the 31st'],
	])('%s reads as "%s"', (text, words) => {
		expect(describeRepeatRule(rule(text))).toBe(words);
	});
});

describe('nextOccurrence — the instant a rule next comes due', () => {
	it('advances a daily rule to the same wall-clock time tomorrow (IST, +05:30)', () => {
		// 09:00 in Asia/Kolkata is 03:30Z. A :30 offset is what stops "add 24
		// hours" from being indistinguishable from "add a calendar day".
		expect(
			next('FREQ=DAILY', '2026-03-02T03:30:00.000Z', '2026-03-02T03:30:00.000Z'),
		).toBe('2026-03-03T03:30:00.000Z');
	});

	it('is strictly after: a rule landing exactly on `after` advances', () => {
		// The defect this guards: returning `after` itself means a reminder that
		// fires at 09:00 is rescheduled to 09:00 — i.e. armed in the past — and the
		// chain stops.
		expect(
			next('FREQ=DAILY', '2026-03-02T03:30:00.000Z', '2026-03-02T03:30:00.000Z'),
		).toBe('2026-03-03T03:30:00.000Z');
	});

	it('is strictly after: one millisecond before the first occurrence returns it', () => {
		expect(
			next('FREQ=DAILY', '2026-03-02T03:30:00.000Z', '2026-03-02T03:29:59.999Z'),
		).toBe('2026-03-02T03:30:00.000Z');
	});

	it('returns the start itself when asked for the next one before it begins', () => {
		expect(
			next('FREQ=WEEKLY;BYDAY=MO', '2026-03-02T03:30:00.000Z', '2026-02-20T00:00:00.000Z'),
		).toBe('2026-03-02T03:30:00.000Z');
	});

	it('skips whole intervals rather than stepping one day at a time', () => {
		// every 3 days from Mon 5 Jan 09:00 IST: Tue 6 Jan is not one.
		expect(
			next('FREQ=DAILY;INTERVAL=3', '2026-01-05T03:30:00.000Z', '2026-01-06T00:00:00.000Z'),
		).toBe('2026-01-08T03:30:00.000Z');
	});

	it('keeps the wall clock across a DST change (Europe/London)', () => {
		// 2026-03-29 is the last Sunday of March: the UK moves to BST and the local
		// offset goes +00:00 → +01:00. A 09:00 reminder stays at 09:00 local, so its
		// UTC instant moves by an hour. Adding 86_400_000 ms would fire it at 10:00.
		expect(
			next('FREQ=DAILY', '2026-03-27T09:00:00.000Z', '2026-03-28T09:00:00.000Z', 'Europe/London'),
		).toBe('2026-03-29T08:00:00.000Z');
		// And the day after the change is 09:00 BST again — 08:00Z.
		expect(
			next('FREQ=DAILY', '2026-03-27T09:00:00.000Z', '2026-03-29T08:00:00.000Z', 'Europe/London'),
		).toBe('2026-03-30T08:00:00.000Z');
	});

	it('handles a rule at 00:30 local without drifting to the previous day', () => {
		const start = '2026-01-05T19:00:00.000Z'; // Tue 6 Jan 00:30 IST
		expect(next('FREQ=DAILY', start, start)).toBe('2026-01-06T19:00:00.000Z');
	});

	it('moves a weekly rule to the next week, across a month boundary', () => {
		// Mon 26 Jan 2026 → Mon 2 Feb 2026.
		expect(
			next('FREQ=WEEKLY;BYDAY=MO', '2026-01-26T03:30:00.000Z', '2026-01-27T00:00:00.000Z'),
		).toBe('2026-02-02T03:30:00.000Z');
	});

	it('holds a weekly rule at its weekday, not at "seven days later"', () => {
		// Asked on Tuesday for the Monday just gone, the next Monday is six days
		// away, not seven — "every Monday" is a weekday, not an interval.
		expect(
			next('FREQ=WEEKLY;BYDAY=MO', '2026-03-02T03:30:00.000Z', '2026-03-03T00:00:00.000Z'),
		).toBe('2026-03-09T03:30:00.000Z');
	});

	it('honours a two-week interval anchored at the first occurrence', () => {
		expect(
			next('FREQ=WEEKLY;INTERVAL=2;BYDAY=MO', '2026-01-05T03:30:00.000Z', '2026-01-06T00:00:00.000Z'),
		).toBe('2026-01-19T03:30:00.000Z');
		// And the week in between is not an occurrence.
		expect(
			next('FREQ=WEEKLY;INTERVAL=2;BYDAY=MO', '2026-01-05T03:30:00.000Z', '2026-01-12T00:00:00.000Z'),
		).toBe('2026-01-19T03:30:00.000Z');
	});

	it('moves a monthly rule to the same day of the next month', () => {
		expect(
			next('FREQ=MONTHLY;BYMONTHDAY=15', '2026-01-15T03:30:00.000Z', '2026-01-20T00:00:00.000Z'),
		).toBe('2026-02-15T03:30:00.000Z');
	});

	it('skips February for a monthly rule on the 31st', () => {
		// The decision, and the reason for it: the phone arms the repeat through
		// Android's `DAY_OF_MONTH_AND_TIME` component, which walks forward a day at a
		// time until the day-of-month matches — so the device itself skips a month
		// that has no 31st. Clamping this function to 28 February would make the
		// server's answer disagree with the alarm actually armed on the phone, and
		// invent an occurrence the user never gets.
		expect(
			next('FREQ=MONTHLY;BYMONTHDAY=31', '2026-01-31T03:30:00.000Z', '2026-02-01T00:00:00.000Z'),
		).toBe('2026-03-31T03:30:00.000Z');
	});

	it('skips February for a monthly rule on the 29th in a non-leap year', () => {
		expect(
			next('FREQ=MONTHLY;BYMONTHDAY=29', '2027-01-29T03:30:00.000Z', '2027-02-01T00:00:00.000Z'),
		).toBe('2027-03-29T03:30:00.000Z');
	});

	it('keeps a monthly rule on the 29th in a leap year', () => {
		// 2028 is a leap year, so February has a 29th and nothing is skipped.
		expect(
			next('FREQ=MONTHLY;BYMONTHDAY=29', '2028-01-29T03:30:00.000Z', '2028-02-01T00:00:00.000Z'),
		).toBe('2028-02-29T03:30:00.000Z');
	});

	it('honours a monthly interval across a year boundary', () => {
		expect(
			next('FREQ=MONTHLY;INTERVAL=3;BYMONTHDAY=10', '2026-10-10T03:30:00.000Z', '2026-11-01T00:00:00.000Z'),
		).toBe('2027-01-10T03:30:00.000Z');
	});

	it("resolves the recurrence in the reminder's own zone, not the server's", () => {
		// One instant, one rule, two zones — and a different next occurrence, because
		// "Tuesday" is a wall-clock fact. 2026-01-05T18:30Z is Tuesday 00:00 in
		// Asia/Kolkata and Monday 13:30 in New York, so the next Tuesday after it is
		// a week away in one zone and tomorrow in the other.
		const start = '2026-01-05T18:30:00.000Z';
		expect(next('FREQ=DAILY', start, start, 'Asia/Kolkata')).toBe('2026-01-06T18:30:00.000Z');
		expect(next('FREQ=DAILY', start, start, 'America/New_York')).toBe('2026-01-06T18:30:00.000Z');
		expect(next('FREQ=WEEKLY;BYDAY=TU', start, start, 'Asia/Kolkata')).toBe(
			'2026-01-12T18:30:00.000Z',
		);
		expect(next('FREQ=WEEKLY;BYDAY=TU', start, start, 'America/New_York')).toBe(
			'2026-01-06T18:30:00.000Z',
		);
	});
});

File: apps/mobile/test/features/reminder_recurrence_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:nova_mobile/core/api/models.dart';
import 'package:nova_mobile/features/reminders/reminder_recurrence.dart';

/// The client half of "remind me every Monday".
///
/// The server stores the rule and the phone has to turn it into the *next*
/// occurrence once the one it was created for is behind it — the row's
/// `trigger_at` is only the first time it goes off, and nothing server-side
/// advances it. This is the pure half of that, and it mirrors
/// `services/api/src/services/reminder-recurrence.ts` case for case, because the
/// two must agree about February.
///
/// The assertions are written against the local wall clock (`.toLocal()`), not
/// against absolute instants, so they hold in whichever zone the test machine
/// happens to be in — "every Monday at 09:00" is a wall-clock promise.
void main() {
  NovaRepeatRule rule(String text) {
    final parsed = NovaRepeatRule.tryParse(text);
    expect(parsed, isNotNull, reason: 'fixture rule $text must parse');
    return parsed!;
  }

  DateTime nextOf(NovaRepeatRule r, DateTime start, DateTime after) {
    final next = nextReminderOccurrence(rule: r, start: start, after: after);
    expect(next, isNotNull, reason: 'expected an occurrence for ${r.value}');
    return next!.toLocal();
  }

  group('NovaRepeatRule.tryParse', () {
    test('parses the three forms the server writes', () {
      expect(
        NovaRepeatRule.tryParse('FREQ=DAILY'),
        isA<NovaRepeatRule>()
            .having((r) => r.frequency, 'frequency', NovaRepeatFrequency.daily)
            .having((r) => r.interval, 'interval', 1),
      );
      expect(
        NovaRepeatRule.tryParse('FREQ=WEEKLY;BYDAY=MO'),
        isA<NovaRepeatRule>()
            .having((r) => r.frequency, 'frequency', NovaRepeatFrequency.weekly)
            .having((r) => r.weekday, 'weekday', DateTime.monday),
      );
      expect(
        NovaRepeatRule.tryParse('FREQ=MONTHLY;BYMONTHDAY=15'),
        isA<NovaRepeatRule>()
            .having((r) => r.frequency, 'frequency', NovaRepeatFrequency.monthly)
            .having((r) => r.dayOfMonth, 'dayOfMonth', 15),
      );
    });

    test('parses an interval, which the REST route can store even though the '
        'assistant does not offer it', () {
      final parsed = NovaRepeatRule.tryParse('FREQ=WEEKLY;INTERVAL=2;BYDAY=SU');
      expect(parsed!.interval, 2);
      expect(parsed.weekday, DateTime.sunday);
    });

    test('is case-insensitive', () {
      expect(NovaRepeatRule.tryParse('freq=weekly;byday=mo')!.value, 'FREQ=WEEKLY;BYDAY=MO');
    });

    test('answers null for anything it cannot interpret, rather than throwing', () {
      // A rule this build does not understand must leave the reminder armed once,
      // not take a sync down. `null` is the safe direction: the reminder still
      // fires at its trigger, it just does not repeat.
      for (final raw in <Object?>[
        null,
        '',
        'every Monday',
        'FREQ=YEARLY',
        'FREQ=WEEKLY',
        'FREQ=DAILY;BYDAY=MO',
        'FREQ=MONTHLY;BYMONTHDAY=32',
      ]) {
        expect(NovaRepeatRule.tryParse(raw), isNull, reason: '$raw must not parse');
      }
    });
  });

  group('NovaReminder parses the server\'s repeatRule', () {
    test('reads `repeatRule` off a real payload', () {
      final reminder = NovaReminder.fromJson(<String, dynamic>{
        'id': 'aaaaaaaa-1111-4111-8111-111111111111',
        'title': 'Take the bins out',
        'triggerAt': '2027-03-01T03:30:00.000Z',
        'timezone': 'Asia/Kolkata',
        'repeatRule': 'FREQ=WEEKLY;BYDAY=MO',
        'dismissed': false,
      });

      expect(reminder.repeatRule, isNotNull);
      expect(reminder.repeatRule!.frequency, NovaRepeatFrequency.weekly);
      expect(reminder.repeatRule!.weekday, DateTime.monday);
    });

    test('leaves an uninterpretable rule as a one-shot instead of failing', () {
      final reminder = NovaReminder.fromJson(<String, dynamic>{
        'id': 'b',
        'title': 'Call the bank',
        'triggerAt': '2027-03-01T03:30:00.000Z',
        'repeatRule': 'whenever',
      });

      expect(reminder.repeatRule, isNull);
      expect(reminder.remindAt, isNotNull);
    });

    test('reads a one-shot reminder exactly as before', () {
      final reminder = NovaReminder.fromJson(<String, dynamic>{
        'id': 'c',
        'title': 'Call the bank',
        'triggerAt': '2027-03-01T03:30:00.000Z',
      });

      expect(reminder.repeatRule, isNull);
    });
  });

  group('nextReminderOccurrence', () {
    test('advances a daily rule to the same wall-clock time tomorrow', () {
      final start = DateTime(2026, 3, 2, 9);
      expect(nextOf(rule('FREQ=DAILY'), start, start), DateTime(2026, 3, 3, 9));
    });

    test('is strictly after: a rule landing exactly on `after` advances', () {
      // Returning `after` itself would re-arm the alarm in the past, and the
      // chain would stop after one occurrence.
      final start = DateTime(2026, 3, 2, 9);
      expect(nextOf(rule('FREQ=DAILY'), start, start), DateTime(2026, 3, 3, 9));
    });

    test('returns the first occurrence when asked before it begins', () {
      final start = DateTime(2026, 3, 2, 9);
      expect(nextOf(rule('FREQ=DAILY'), start, DateTime(2026, 2, 20)), start);
    });

    test('skips whole intervals instead of stepping a day at a time', () {
      final start = DateTime(2026, 1, 5, 9);
      expect(
        nextOf(rule('FREQ=DAILY;INTERVAL=3'), start, DateTime(2026, 1, 6)),
        DateTime(2026, 1, 8, 9),
      );
    });

    test('moves a weekly rule to its weekday across a month boundary', () {
      // Monday 26 January 2026 -> Monday 2 February 2026.
      final start = DateTime(2026, 1, 26, 9);
      expect(start.weekday, DateTime.monday, reason: 'the fixture must be a Monday');
      final next = nextOf(rule('FREQ=WEEKLY;BYDAY=MO'), start, DateTime(2026, 1, 27));
      expect(next, DateTime(2026, 2, 2, 9));
      expect(next.weekday, DateTime.monday);
    });

    test('holds a weekly rule at its weekday, not at "seven days later"', () {
      final start = DateTime(2026, 3, 2, 9); // Monday
      expect(nextOf(rule('FREQ=WEEKLY;BYDAY=MO'), start, DateTime(2026, 3, 3)), DateTime(2026, 3, 9, 9));
    });

    test('honours a two-week interval', () {
      final start = DateTime(2026, 1, 5, 9); // Monday
      expect(
        nextOf(rule('FREQ=WEEKLY;INTERVAL=2;BYDAY=MO'), start, DateTime(2026, 1, 12)),
        DateTime(2026, 1, 19, 9),
      );
    });

    test('moves a monthly rule to the same day of the next month', () {
      final start = DateTime(2026, 1, 15, 9);
      expect(
        nextOf(rule('FREQ=MONTHLY;BYMONTHDAY=15'), start, DateTime(2026, 1, 20)),
        DateTime(2026, 2, 15, 9),
      );
    });

    test('skips February for a monthly rule on the 31st', () {
      // The decision, matched to the OS: Android's DAY_OF_MONTH_AND_TIME repeat
      // walks forward a day at a time until the day-of-month matches, so the phone
      // skips February for a 31st rule. Clamping here to 28 February would arm an
      // alarm the phone will never fire.
      final start = DateTime(2026, 1, 31, 9);
      final next = nextOf(rule('FREQ=MONTHLY;BYMONTHDAY=31'), start, DateTime(2026, 2, 1));
      expect(next.month, 3);
      expect(next.day, 31);
      expect(next.hour, 9);
    });

    test('skips February for the 29th in a non-leap year, and keeps it in a leap year', () {
      final nonLeap = nextOf(
        rule('FREQ=MONTHLY;BYMONTHDAY=29'),
        DateTime(2027, 1, 29, 9),
        DateTime(2027, 2, 1),
      );
      expect(nonLeap, DateTime(2027, 3, 29, 9));

      final leap = nextOf(
        rule('FREQ=MONTHLY;BYMONTHDAY=29'),
        DateTime(2028, 1, 29, 9),
        DateTime(2028, 2, 1),
      );
      expect(leap, DateTime(2028, 2, 29, 9));
    });

    test('honours a monthly interval across a year boundary', () {
      final start = DateTime(2026, 10, 10, 9);
      expect(
        nextOf(rule('FREQ=MONTHLY;INTERVAL=3;BYMONTHDAY=10'), start, DateTime(2026, 11, 1)),
        DateTime(2027, 1, 10, 9),
      );
    });

    test('preserves the time of day of an instant parsed from an ISO payload', () {
      // `_parseDate` already converts to local, so the hours the server sent are
      // the ones the user was shown. If this ever regressed to reading UTC fields,
      // a 09:00 reminder would be armed at 03:30 on a device in IST.
      final start = DateTime.parse('2026-03-02T03:30:00.000Z').toLocal();
      final next = nextOf(rule('FREQ=DAILY'), start, start);
      expect(next.hour, start.hour);
      expect(next.minute, start.minute);
      expect(
        next,
        DateTime(start.year, start.month, start.day + 1, start.hour, start.minute, start.second),
      );
    });
  });
}
